mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -0,0 +1,1758 @@
+  if (!func) {
+    OpBuilder builder(module.getContext());
+    func = FuncOp::create(builder.getUnknownLoc(), funcName,
+                          builder.getFunctionType(inputs, results));
+    func.sym_visibilityAttr(builder.getStringAttr("private"));
+    module.push_back(func);
+  }
+
+  return func;
+}
+
+static mlir::Value makeNDMemRef(OpBuilder &b, mlir::Value var, uint32_t ndim) {
+  auto context = b.getContext();
+  auto oprType = var.getType().template cast<ShapedType>();
+  if (!oprType.hasStaticShape())
+    return mlir::Value();
+
+  auto shape = oprType.getShape();
+  auto loc = var.getLoc();
+
+  if (shape.size() > ndim) {
+    // Collapse last dims
+    SmallVector<int64_t, 5> colShape;
+    SmallVector<ReassociationExprs, 5> reassocs;
+    uint32_t dim = 0;
+    for (; dim < ndim - 1; ++dim) {
+      colShape.push_back(shape[dim]);
+      reassocs.push_back({getAffineDimExpr(dim, context)});
+    }
+
+    // Last dim
+    uint64_t lastDim = 1;
+    SmallVector<AffineExpr, 2> exprs;
+    for (; dim < shape.size(); ++dim) {
+      lastDim *= shape[dim];
+      exprs.push_back(getAffineDimExpr(dim, context));
+    }
+    colShape.push_back(lastDim);
+    reassocs.push_back(exprs);
+
+    auto colType = MemRefType::get(colShape, oprType.getElementType());
+
+    // Emit memref.collapse_shape
+    var = b.create<memref::CollapseShapeOp>(loc, colType, var, reassocs);
+  } else if (shape.size() < ndim) {
+    // Expand last dims
+    SmallVector<int64_t, 5> expShape;
+    SmallVector<ReassociationExprs, 5> reassocs;
+    uint32_t dim = 0;
+    for (; dim < shape.size() - 1; ++dim) {
+      expShape.push_back(shape[dim]);
+      reassocs.push_back({getAffineDimExpr(dim, context)});
+    }
+
+    // Last dim
+    expShape.push_back(shape[dim]);
+    SmallVector<AffineExpr, 2> exprs;
+    for (; dim < ndim; ++dim) {
+      expShape.push_back(1);
+      exprs.push_back(getAffineDimExpr(dim, context));
+    }
+    expShape.pop_back();
+    reassocs.push_back(exprs);
+
+    auto expType = MemRefType::get(expShape, oprType.getElementType());
+
+    // Emit memref.collapse_shape
+    var = b.create<memref::ExpandShapeOp>(loc, expType, var, reassocs);
+  }
+
+  return var;
+}
+
+static FuncOp createGPUWrapper(ModuleOp &module, const KernelIF &kernel) {
+  auto context = module.getContext();
+  OpBuilder b(context);
+  auto loc = kernel.func->getLoc();
+
+  // Create gpu wrapper function
+  auto kfunc = kernel.func;
+  std::string funcName = kfunc.getName().str() + "_gpu";
+  auto gpuWrapperFuncType = b.getFunctionType(kernel.params, {});
+
+  auto gpuWrapperFunc =
+      FuncOp::create(loc, StringRef(funcName), gpuWrapperFuncType);
+  module.push_back(gpuWrapperFunc);
+
+  // Emit gpu convolution logic.
+  auto block = gpuWrapperFunc.addEntryBlock();
+  b.setInsertionPoint(block, block->begin());
+
+  // Emit device selection
+  if (deviceNum.getNumOccurrences() > 0)
+    b.create<gpu::SetDefaultDeviceOp>(
```

Comment:
Note to future us (not for this PR but): add an option that'll query the environment for how many GPUs there are and pick one in some reasonably fair way so that we can spread the load out over multiple GPUs